<a href="https://colab.research.google.com/github/sandeepbansode/Basic_Python/blob/main/17_Basics_Python_Pandas_DateTime.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Working with date and time using Pandas**

In [4]:
import pandas as pd
data = pd.date_range('10/05/2022' , periods = 10 , freq = 'H')

data

DatetimeIndex(['2022-10-05 00:00:00', '2022-10-05 01:00:00',
               '2022-10-05 02:00:00', '2022-10-05 03:00:00',
               '2022-10-05 04:00:00', '2022-10-05 05:00:00',
               '2022-10-05 06:00:00', '2022-10-05 07:00:00',
               '2022-10-05 08:00:00', '2022-10-05 09:00:00'],
              dtype='datetime64[ns]', freq='H')

Datetime features can be divided into two categories. The first one time moments in a period and second the time passed since a particular period. These features can be very useful to understand the patterns in the data.

Divide a given date into features –

pandas.Series.dt.year returns the year of the date time.

pandas.Series.dt.month returns the month of the date time.

pandas.Series.dt.day returns the day of the date time.

pandas.Series.dt.hour returns the hour of the date time.

pandas.Series.dt.minute returns the minute of the date time.

In [8]:
# Create date and time with dataframe

datte = pd.DataFrame()

datte["date"] = pd.date_range('20/05/2022', periods = 10, freq = 'H')

print(datte)

                 date
0 2022-05-20 00:00:00
1 2022-05-20 01:00:00
2 2022-05-20 02:00:00
3 2022-05-20 03:00:00
4 2022-05-20 04:00:00
5 2022-05-20 05:00:00
6 2022-05-20 06:00:00
7 2022-05-20 07:00:00
8 2022-05-20 08:00:00
9 2022-05-20 09:00:00


In [15]:
import pandas as pd

# Create features for year, month, day, hour, and minute
datte['year'] = datte['date'].dt.year
datte['month'] = datte['date'].dt.month
datte['day'] = datte['date'].dt.day
datte['hour'] = datte['date'].dt.hour
datte['minute']= datte['date'].dt.minute

In [16]:
datte.head()

date  year  month  day  hour  minute
0 2022-05-20 00:00:00  2022      5   20     0       0
1 2022-05-20 01:00:00  2022      5   20     1       0
2 2022-05-20 02:00:00  2022      5   20     2       0
3 2022-05-20 03:00:00  2022      5   20     3       0
4 2022-05-20 04:00:00  2022      5   20     4       0

In [23]:
"""Code #4: To get the present time"""


from datetime import *


today = date.today()

print("TOday's date :",today )

TOday's date : 2022-05-10


In [42]:
#Example 2: Current date in different formats

from datetime import *

today = date.today()


#dd/mm/yyyy

date1 = today.strftime('%d-%-m%-y')

print(date1)

#mm/dd/yyyy

date2 = today.strftime("%m-%d-%y")

print(date2)


#yyyy/mm/dd

date3 = today.strftime("%m-%d-%y")

print(date3)

date4 = today.strftime("%m-%d-%Y")

print(date4)

date5 = today.strftime("%M-%D-%Y")

print(date5)

date6 = today.strftime("%b, %d,%Y")
print(date6)


date7 = today.strftime("%D")

print(date7)


date8 = today.strftime("%M")
print(date8)

date9 = today.strftime("%Y")

print(date9)

10-522
05-10-22
05-10-22
05-10-2022
00-05/10/22-2022
May, 10,2022
05/10/22
00
2022


# **Timedelta**
So far, we have seen how to create a DateTime object and how to format it. But sometimes, you might have to find the duration between two dates, which can be another very useful feature that you can derive from a dataset. This duration is, however, returned as a timedelta object.

In [45]:
from datetime import *

d1 = datetime(2022,4,15,12,5,10)
d2 = datetime(2020,4,16,15,6,10)
duration = d1-d2
print(d1)
print(d2)
print(type(duration))
duration

2022-04-15 12:05:10
2020-04-16 15:06:10
<class 'datetime.timedelta'>


datetime.timedelta(days=728, seconds=75540)

# **Datetime and Pandas Dataframe**

In [46]:
import pandas as pd

url = 'http://bit.ly/uforeports'

# read csv file
df = pd.read_csv(url)		
df.head()

City Colors Reported Shape Reported State             Time
0                Ithaca             NaN       TRIANGLE    NY   6/1/1930 22:00
1           Willingboro             NaN          OTHER    NJ  6/30/1930 20:00
2               Holyoke             NaN           OVAL    CO  2/15/1931 14:00
3               Abilene             NaN           DISK    KS   6/1/1931 13:00
4  New York Worlds Fair             NaN          LIGHT    NY  4/18/1933 19:00

In [47]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18241 entries, 0 to 18240
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   City             18216 non-null  object
 1   Colors Reported  2882 non-null   object
 2   Shape Reported   15597 non-null  object
 3   State            18241 non-null  object
 4   Time             18241 non-null  object
dtypes: object(5)
memory usage: 712.7+ KB


In [48]:
df.tail()

City Colors Reported Shape Reported State              Time
18236   Grant Park             NaN       TRIANGLE    IL  12/31/2000 23:00
18237  Spirit Lake             NaN           DISK    IA  12/31/2000 23:00
18238  Eagle River             NaN            NaN    WI  12/31/2000 23:45
18239  Eagle River             RED          LIGHT    WI  12/31/2000 23:45
18240         Ybor             NaN           OVAL    FL  12/31/2000 23:59

In [50]:
df.sample()

City Colors Reported Shape Reported State            Time
3558  Waddy             NaN          LIGHT    KY  7/3/1977 11:00

In [51]:
df.sample(5)

City Colors Reported Shape Reported State  \
16933                             Aurora             NaN         SPHERE    OR   
17221                           Portland             RED          OTHER    OR   
16923                             Tooele             NaN            NaN    UT   
11975  Blewett Pass-top of mountain pass             NaN          LIGHT    WA   
6950                            Columbus             NaN           DISK    OH   

                   Time  
16933   7/15/2000 16:00  
17221    8/16/2000 2:45  
16923   7/14/2000 22:30  
11975    8/14/1998 0:00  
6950   11/27/1992 19:50

In [52]:
# Convert the Time column to datetime format

df["Time"] = pd.to_datetime(df.Time)

In [53]:
df.head()

City Colors Reported Shape Reported State  \
0                Ithaca             NaN       TRIANGLE    NY   
1           Willingboro             NaN          OTHER    NJ   
2               Holyoke             NaN           OVAL    CO   
3               Abilene             NaN           DISK    KS   
4  New York Worlds Fair             NaN          LIGHT    NY   

                 Time  
0 1930-06-01 22:00:00  
1 1930-06-30 20:00:00  
2 1931-02-15 14:00:00  
3 1931-06-01 13:00:00  
4 1933-04-18 19:00:00

In [54]:
df.tail()

City Colors Reported Shape Reported State                Time
18236   Grant Park             NaN       TRIANGLE    IL 2000-12-31 23:00:00
18237  Spirit Lake             NaN           DISK    IA 2000-12-31 23:00:00
18238  Eagle River             NaN            NaN    WI 2000-12-31 23:45:00
18239  Eagle River             RED          LIGHT    WI 2000-12-31 23:45:00
18240         Ybor             NaN           OVAL    FL 2000-12-31 23:59:00

In [56]:
df.sample(4)

City Colors Reported Shape Reported State  \
12478           Fort Worth             NaN          LIGHT    TX   
9203             Arlington             NaN           DISK    WA   
13166  South San Francisco             NaN           DISK    CA   
10353               Golden             NaN          LIGHT    CO   

                     Time  
12478 1998-11-01 15:00:00  
9203  1996-04-05 23:30:00  
13166 1999-03-11 15:00:00  
10353 1997-06-10 22:30:00

In [57]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18241 entries, 0 to 18240
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   City             18216 non-null  object        
 1   Colors Reported  2882 non-null   object        
 2   Shape Reported   15597 non-null  object        
 3   State            18241 non-null  object        
 4   Time             18241 non-null  datetime64[ns]
dtypes: datetime64[ns](1), object(4)
memory usage: 712.7+ KB


# **Day first format**
By default, to_datetime() will parse string with month first (MM/DD, MM DD, or MM-DD) format, and this arrangement is relatively unique in the United State.

In most of the rest of the world, the day is written first (DD/MM, DD MM, or DD-MM). If you would like Pandas to consider day first instead of month, you can set the argument dayfirst to True.

In [67]:
df = pd.DataFrame({'date':['10-12-2022','12-12-2022','15-12-2022'],
                   'values':[1,2,3]})

print(df.info())

print('\n\n\n',df)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   date    3 non-null      object
 1   values  3 non-null      int64 
dtypes: int64(1), object(1)
memory usage: 176.0+ bytes
None



          date  values
0  10-12-2022       1
1  12-12-2022       2
2  15-12-2022       3


In [76]:
df1 = df['date']

df['date'] = pd.to_datetime(df.date)

print(df.info())

print("**********************************")

print(df1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   date    3 non-null      datetime64[ns]
 1   value   3 non-null      int64         
dtypes: datetime64[ns](1), int64(1)
memory usage: 176.0 bytes
None
**********************************
0   2016-10-06 20:30:00
1   2016-01-07 19:45:30
2   2013-12-10 04:05:01
Name: date, dtype: datetime64[ns]


In [66]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   date    3 non-null      object
 1   values  3 non-null      int64 
dtypes: int64(1), object(1)
memory usage: 176.0+ bytes



# **Custome format**
By default, strings are parsed using the Pandas built-in parser from dateutil.parser.parse. Sometimes, your strings might be in a custom format, for example, YYYY-DD-MM HH:MM:SS. Pandas to_datetime() has an argument called format that allows you to pass a custom format:

In [68]:
df = pd.DataFrame({'date': ['2016-6-10 20:30:0', 
                            '2016-7-1 19:45:30', 
                            '2013-10-12 4:5:1'],
                   'value': [2, 3, 4]})
print(df)
print("*****************************")
df['date'] = pd.to_datetime(df['date'], format="%Y-%d-%m %H:%M:%S")
df

                date  value
0  2016-6-10 20:30:0      2
1  2016-7-1 19:45:30      3
2   2013-10-12 4:5:1      4
*****************************


date  value
0 2016-10-06 20:30:00      2
1 2016-01-07 19:45:30      3
2 2013-12-10 04:05:01      4

In [69]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   date    3 non-null      datetime64[ns]
 1   value   3 non-null      int64         
dtypes: datetime64[ns](1), int64(1)
memory usage: 176.0 bytes


# **Pandas Timestamp.replace**
Pandas Timestamp.replace() function is used to replace the member values of the given Timestamp. The function implements datetime.replace, and it also handles nanoseconds.

In [85]:
import pandas as pd

tiss = pd.Timestamp(year = 2022, month = 6,day = 21 , hour = 15 , second = 45 , tz= 'US/Central')


print(tiss)

2022-06-21 15:00:45-05:00


In [88]:

tiss = pd.Timestamp(year = 2022, month = 6,day = 21 , hour = 15 , second = 45 , tz= 'Asia/Kolkata')

print(tiss)

2022-06-21 15:00:45+05:30


In [89]:
tiss.replace(year = 2019)


Timestamp('2019-06-21 15:00:45+0530', tz='Asia/Kolkata')

In [90]:
#Example #2: Use Timestamp.replace() function to replace the year, month and hour value in the given Timestamp.

# importing pandas as pd
import pandas as pd

# Create the Timestamp object
ts = pd.Timestamp(year = 2009, month = 5, day = 31,
				hour = 4, second = 49, tz = 'Europe/Berlin')

# Print the Timestamp object
print(ts)

2009-05-31 04:00:49+02:00


In [92]:
ts.replace(year = 2022 , month = 12,hour = 20)

Timestamp('2022-12-31 20:00:49+0100', tz='Europe/Berlin')

# **pandas.date_range() method**
pandas.date_range() is one of the general functions in Pandas which is used to return a fixed frequency DatetimeIndex.

In [125]:
import pandas as pd

date1 = pd.date_range(start = '1-2-2015',end = '1-10-2015',freq='10H')

In [126]:
print(date1)

DatetimeIndex(['2015-01-02 00:00:00', '2015-01-02 10:00:00',
               '2015-01-02 20:00:00', '2015-01-03 06:00:00',
               '2015-01-03 16:00:00', '2015-01-04 02:00:00',
               '2015-01-04 12:00:00', '2015-01-04 22:00:00',
               '2015-01-05 08:00:00', '2015-01-05 18:00:00',
               '2015-01-06 04:00:00', '2015-01-06 14:00:00',
               '2015-01-07 00:00:00', '2015-01-07 10:00:00',
               '2015-01-07 20:00:00', '2015-01-08 06:00:00',
               '2015-01-08 16:00:00', '2015-01-09 02:00:00',
               '2015-01-09 12:00:00', '2015-01-09 22:00:00'],
              dtype='datetime64[ns]', freq='10H')


In [127]:
df = pd.DataFrame(date1)

In [128]:
df.head()

0
0 2015-01-02 00:00:00
1 2015-01-02 10:00:00
2 2015-01-02 20:00:00
3 2015-01-03 06:00:00
4 2015-01-03 16:00:00

In [129]:
df.sample()

0
0 2015-01-02

In [130]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   0       20 non-null     datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 288.0 bytes


In [131]:
import pandas as pd

per1 = pd.date_range(start ='1-1-2018',
		end ='1-05-2018', freq ='5H')

for val in per1:
	print(val)

2018-01-01 00:00:00
2018-01-01 05:00:00
2018-01-01 10:00:00
2018-01-01 15:00:00
2018-01-01 20:00:00
2018-01-02 01:00:00
2018-01-02 06:00:00
2018-01-02 11:00:00
2018-01-02 16:00:00
2018-01-02 21:00:00
2018-01-03 02:00:00
2018-01-03 07:00:00
2018-01-03 12:00:00
2018-01-03 17:00:00
2018-01-03 22:00:00
2018-01-04 03:00:00
2018-01-04 08:00:00
2018-01-04 13:00:00
2018-01-04 18:00:00
2018-01-04 23:00:00


In [132]:
# importing pandas as pd
import pandas as pd

daan1 = pd.date_range(start ='1-1-2018',
		end ='8-01-2018', freq ='M')

daan2 = pd.date_range(start ='1-1-2018',
		end ='11-01-2018', freq ='3M')

print(daan1, '\n\n', daan2)

DatetimeIndex(['2018-01-31', '2018-02-28', '2018-03-31', '2018-04-30',
               '2018-05-31', '2018-06-30', '2018-07-31'],
              dtype='datetime64[ns]', freq='M') 

 DatetimeIndex(['2018-01-31', '2018-04-30', '2018-07-31', '2018-10-31'], dtype='datetime64[ns]', freq='3M')


# **Usecase**

In [173]:
url='https://raw.githubusercontent.com/jenfly/opsd/master/opsd_germany_daily.csv'
data = pd.read_csv(url,sep=",")

In [174]:
url

'https://raw.githubusercontent.com/jenfly/opsd/master/opsd_germany_daily.csv'

In [175]:
data.head()

Date  Consumption  Wind  Solar  Wind+Solar
0  2006-01-01     1069.184   NaN    NaN         NaN
1  2006-01-02     1380.521   NaN    NaN         NaN
2  2006-01-03     1442.533   NaN    NaN         NaN
3  2006-01-04     1457.217   NaN    NaN         NaN
4  2006-01-05     1477.131   NaN    NaN         NaN


# **Converting data to correct format**
when working with time series, it becomes much easier if we have the Datecolumn represented as a Timestamp. Timestamp is the main pandas data structures for working with dates and times. The pandas function to_datetime() can help us convert a string to a proper date/time format

In [176]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4383 entries, 0 to 4382
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Date         4383 non-null   object 
 1   Consumption  4383 non-null   float64
 2   Wind         2920 non-null   float64
 3   Solar        2188 non-null   float64
 4   Wind+Solar   2187 non-null   float64
dtypes: float64(4), object(1)
memory usage: 171.3+ KB


In [177]:
data['Date'] = pd.to_datetime(data['Date'])   #pd.to_datetime(data.Date)  data['Date'] = pd.to_datetime(data['Date']) 

In [178]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4383 entries, 0 to 4382
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         4383 non-null   datetime64[ns]
 1   Consumption  4383 non-null   float64       
 2   Wind         2920 non-null   float64       
 3   Solar        2188 non-null   float64       
 4   Wind+Solar   2187 non-null   float64       
dtypes: datetime64[ns](1), float64(4)
memory usage: 171.3 KB


In [179]:
data.head()

Date  Consumption  Wind  Solar  Wind+Solar
0 2006-01-01     1069.184   NaN    NaN         NaN
1 2006-01-02     1380.521   NaN    NaN         NaN
2 2006-01-03     1442.533   NaN    NaN         NaN
3 2006-01-04     1457.217   NaN    NaN         NaN
4 2006-01-05     1477.131   NaN    NaN         NaN

In [180]:
data.dtypes

Date           datetime64[ns]
Consumption           float64
Wind                  float64
Solar                 float64
Wind+Solar            float64
dtype: object

We will now go ahead and set this column as the index for the dataframe using the set_index() call.

In [181]:
data = data.set_index('Date')

data

Consumption     Wind   Solar  Wind+Solar
Date                                                
2006-01-01   1069.18400      NaN     NaN         NaN
2006-01-02   1380.52100      NaN     NaN         NaN
2006-01-03   1442.53300      NaN     NaN         NaN
2006-01-04   1457.21700      NaN     NaN         NaN
2006-01-05   1477.13100      NaN     NaN         NaN
...                 ...      ...     ...         ...
2017-12-27   1263.94091  394.507  16.530     411.037
2017-12-28   1299.86398  506.424  14.162     520.586
2017-12-29   1295.08753  584.277  29.854     614.131
2017-12-30   1215.44897  721.247   7.467     728.714
2017-12-31   1107.11488  721.176  19.980     741.156

[4383 rows x 4 columns]

In [182]:
data1.index

DatetimeIndex(['2006-01-01', '2006-01-02', '2006-01-03', '2006-01-04',
               '2006-01-05', '2006-01-06', '2006-01-07', '2006-01-08',
               '2006-01-09', '2006-01-10',
               ...
               '2017-12-22', '2017-12-23', '2017-12-24', '2017-12-25',
               '2017-12-26', '2017-12-27', '2017-12-28', '2017-12-29',
               '2017-12-30', '2017-12-31'],
              dtype='datetime64[ns]', name='Date', length=4383, freq=None)

An interesting thing to note here is thatfreq = None. What this means is that it is not known if data is collected by the hour, by day, my minute, etc. However, just by eyeballing the indices, we can see that it looks like the data was collected by the day. It would be nice to explicitly put this info into the data frame as well and we will be seeing how to do this shortly! But first a quick detour…

# What if I want to set Date and Time, both as the index?
You will come across datasets where the Date and Time were recorded as separate columns at the time of data collection. A simple yet neat trick to set them as data index is:

Concatenate the two columns but with a space between them.

The space is important!!

Convert this concatenated column to a Timestamp using to_datetime().

Set this concatenated column as the index using set_index().

In [147]:
df['Datetime'] = pd.to_datetime(df['date'] + ' ' + df['time'])
df = df.set_index('Datetime')

# Missing value imputations
As we mentioned previously, quick glance at the data suggests it was collected at an interval of 24 hours (or a day). However, the data reflects freq = None. We can correct it as follows:

In [183]:
data_freq = data.asfreq('D')
data_freq

Consumption     Wind   Solar  Wind+Solar
Date                                                
2006-01-01   1069.18400      NaN     NaN         NaN
2006-01-02   1380.52100      NaN     NaN         NaN
2006-01-03   1442.53300      NaN     NaN         NaN
2006-01-04   1457.21700      NaN     NaN         NaN
2006-01-05   1477.13100      NaN     NaN         NaN
...                 ...      ...     ...         ...
2017-12-27   1263.94091  394.507  16.530     411.037
2017-12-28   1299.86398  506.424  14.162     520.586
2017-12-29   1295.08753  584.277  29.854     614.131
2017-12-30   1215.44897  721.247   7.467     728.714
2017-12-31   1107.11488  721.176  19.980     741.156

[4383 rows x 4 columns]

In [152]:
data

Date  Consumption     Wind   Solar  Wind+Solar
0    2006-01-01   1069.18400      NaN     NaN         NaN
1    2006-01-02   1380.52100      NaN     NaN         NaN
2    2006-01-03   1442.53300      NaN     NaN         NaN
3    2006-01-04   1457.21700      NaN     NaN         NaN
4    2006-01-05   1477.13100      NaN     NaN         NaN
...         ...          ...      ...     ...         ...
4378 2017-12-27   1263.94091  394.507  16.530     411.037
4379 2017-12-28   1299.86398  506.424  14.162     520.586
4380 2017-12-29   1295.08753  584.277  29.854     614.131
4381 2017-12-30   1215.44897  721.247   7.467     728.714
4382 2017-12-31   1107.11488  721.176  19.980     741.156

[4383 rows x 5 columns]

Note: Available frequencies in pandas include hourly (‘H’), calendar daily (‘D’), business daily (‘B’), weekly (‘W’), monthly (‘M’), quarterly (‘Q’), annual (‘A’), and many others.

What we have done above is say “hey, the data was collected at a day’s interval and hence each row is a new day”. So technically, we should have values for power consumption, solar production, wind production, etc for all days from 2006–2017.

In case, some days are missing in our data, the above code is going to insert empty rows and in each of these rows the values corresponding to the columns would be allNaNs. In order to avoid these NaN values, we can tell as.freq function how to fill these null values

In [184]:
data_freq = data.asfreq('D', method = 'ffill')
data_freq

Consumption     Wind   Solar  Wind+Solar
Date                                                
2006-01-01   1069.18400      NaN     NaN         NaN
2006-01-02   1380.52100      NaN     NaN         NaN
2006-01-03   1442.53300      NaN     NaN         NaN
2006-01-04   1457.21700      NaN     NaN         NaN
2006-01-05   1477.13100      NaN     NaN         NaN
...                 ...      ...     ...         ...
2017-12-27   1263.94091  394.507  16.530     411.037
2017-12-28   1299.86398  506.424  14.162     520.586
2017-12-29   1295.08753  584.277  29.854     614.131
2017-12-30   1215.44897  721.247   7.467     728.714
2017-12-31   1107.11488  721.176  19.980     741.156

[4383 rows x 4 columns]

ffill refers to forward fill. This means when a null is encountered for a particular column, it will be replaced by the value in the previous row. As an example:

# **Resampling**
Resampling simply refers to an aggregation of data over a certain time period. Its performance is similar to the group by function in SQL, i.e. data is first split into time bins and some computation is performed on each bin. For instance, given our daily data, we can resample (or bin) it by the month or year and generate some relevant statistics such as minimum, maximum, or mean values.

# Weekly Resampling
To compute weekly mean values for electricity consumption, wind, and solar production:

In [185]:
data_columns = ['Consumption', 'Wind', 'Solar', 'Wind+Solar']
data_weekly_mean = data[data_columns].resample('W').mean() # W stands for weekly
data_weekly_mean

Consumption        Wind      Solar  Wind+Solar
Date                                                      
2006-01-01  1069.184000         NaN        NaN         NaN
2006-01-08  1381.300143         NaN        NaN         NaN
2006-01-15  1486.730286         NaN        NaN         NaN
2006-01-22  1490.031143         NaN        NaN         NaN
2006-01-29  1514.176857         NaN        NaN         NaN
...                 ...         ...        ...         ...
2017-12-03  1536.236314  284.334286  18.320857  302.655143
2017-12-10  1554.824946  636.514714  16.440286  652.955000
2017-12-17  1543.856889  442.531857  18.143714  460.675571
2017-12-24  1440.342401  339.018429   9.895143  348.913571
2017-12-31  1203.265211  604.699143  19.240143  623.939286

[627 rows x 4 columns]

If you notice the indices in the output, you would observe that they are at a gap of one week, we have Jan1 2006 followed by Jan8 2006, and so on. The first row above, labeled 2006–01–01, contains the mean of all the data contained in the time bin 2006–01–01 through 2006–01–07.

# Monthly Resampling
Similarly, to compute monthly maximum values for electricity consumption, wind, and solar production:

In [186]:
data_columns = ['Consumption', 'Wind', 'Solar', 'Wind+Solar']
data_monthly_max = data[data_columns].resample('M').max() # W stands for weekly
data_monthly_max

Consumption     Wind    Solar  Wind+Solar
Date                                                 
2006-01-31   1613.31200      NaN      NaN         NaN
2006-02-28   1606.66700      NaN      NaN         NaN
2006-03-31   1545.83400      NaN      NaN         NaN
2006-04-30   1447.79800      NaN      NaN         NaN
2006-05-31   1380.17500      NaN      NaN         NaN
...                 ...      ...      ...         ...
2017-08-31   1434.05200  455.379  216.117     601.010
2017-09-30   1460.61621  761.751  168.123     838.572
2017-10-31   1508.51484  819.818  123.715     846.617
2017-11-30   1617.03309  743.280   57.358     772.775
2017-12-31   1651.90418  812.422   42.994     822.371

[144 rows x 4 columns]

# Rolling windows
This is quite similar to the resampling process that we just learned. The difference is that the bins over which some aggregating functions are performed) are overlapping.

Bins in case of weekly resampling: Jan1- Jan 7; Jan8 - Jan14, Jan 15 - Jan 21, etc

Bins in case of weekly rolling: Jan1- Jan7; Jan 2- Jan 8, Jan 3- Jan 9, etc.

To compute a 7-day rolling mean:

In [187]:
data_columns = ['Consumption', 'Wind', 'Solar', 'Wind+Solar']
data_7d_rol = data[data_columns].rolling(window = 7, center = True).mean()
data_7d_rol

Consumption        Wind      Solar  Wind+Solar
Date                                                      
2006-01-01          NaN         NaN        NaN         NaN
2006-01-02          NaN         NaN        NaN         NaN
2006-01-03          NaN         NaN        NaN         NaN
2006-01-04  1361.471429         NaN        NaN         NaN
2006-01-05  1381.300143         NaN        NaN         NaN
...                 ...         ...        ...         ...
2017-12-27  1208.214129  617.734286  17.807143  635.541429
2017-12-28  1203.265211  604.699143  19.240143  623.939286
2017-12-29          NaN         NaN        NaN         NaN
2017-12-30          NaN         NaN        NaN         NaN
2017-12-31          NaN         NaN        NaN         NaN

[4383 rows x 4 columns]

In the above command, center = True means for the time bin, say Jan 1 to Jan 8, the rolling mean would be calculated and placed next to the center of the bin i.e. Jan 4. To make it clearer, let's check the output of the above code:

As you can see, the mean consumption value (of 1361.471) in 2006–01–04 was calculated by averaging the values from 2006–01–01 to 2006–01–07. Similarly, the mean consumption value (of 1381.300) in 2006–01–05 was calculated by averaging the values from 2006–01–02 to 2006–01–08.

By the same logic, to calculate the mean consumption value for 2006–01–01, we would need to average the values from 2005–12–29 to 2006–01–04. However, the data for the year 2005 is missing and thus we obtain a Null for the first few rows.

For more info: https://towardsdatascience.com/time-series-analysis-using-pandas-in-python-f726d87a97d8